### PR TITLE
Enables remotes submodules when cloning a Git repository (obviously w…

### PR DESCRIPTION
**Pull request by MatthieuGonnet**
…hen using submodules).

## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the `master/buildbot/newsfragments` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

**Comment by MatthieuGonnet** [comment]
Hi @tardyp !
Is there a repository with submodules in Buildbot's infra to write unit tests on ?

**Comment by tardyp** [comment]
Hi.
We have unit test that compare command line generated so we don't actually
require real servers for tests

**Comment by MatthieuGonnet** [comment]
Smart !
I will fix my PR !
Thx.

**Comment by tardyp** [comment]
Hi, do you still plan to finish this PR?
would be a great addition to buildbot

**Comment by MatthieuGonnet** [comment]
> Hi, do you still plan to finish this PR?
> would be a great addition to buildbot

Yes, sorry I was a bit busy. I will keep you in touch.
Regards

**Comment by MatthieuGonnet** [comment]
I am sorry for how long it was. I finally got some time "thanks to" the containement. Let me know if you need something else.

File: master/buildbot/steps/source/git.py [hunk 1]
```diff
@@ -45,35 +45,64 @@ def isTrueOrIsExactlyZero(v):
 
 git_describe_flags = [
     # on or off
-    ('all', lambda v: ['--all'] if v else None),
```
Comment (p12tic) [inline master/buildbot/steps/source/git.py:48]: This PR is almost unreviewable due to all the extra changes. Can they be reverted?

File: master/buildbot/steps/source/git.py [hunk 2]
```diff
@@ -45,35 +45,64 @@ def isTrueOrIsExactlyZero(v):
...
+        "repourl",
+        "reference",
+        "branch",
+        "codebase",
+        "mode",
+        "method",
+        "origin",
+    ]
+
+    def __init__(
```
Comment (p12tic) [inline master/buildbot/steps/source/git.py:85]: Could we keep the style in line with the rest of the codebase? That is, indenting at the opening parentheses.

File: master/buildbot/steps/source/git.py [hunk 3]
```diff
@@ -427,15 +473,16 @@ def _fullClone(self, shallowClone=False):
...
+            res = yield self._dovccmd(
+                ["reset", "--hard", self.revision, "--"], shallowClone
+            )
         # init and update submodules, recursively. If there's not recursion
         # it will not do it.
         if self.submodules:
-            res = yield self._dovccmd(['submodule', 'update',
-                                       '--init', '--recursive'],
-                                      shallowClone)
+            cmdArgs = ["submodule", "update", "--init", "--recursive"]
```
Comment (p12tic) [inline master/buildbot/steps/source/git.py:482]: snake_case is preferred for local variables.

**Comment by MatthieuGonnet** [comment]
> I think the idea is good in principle, but around 98% of the PR are unnecessary changes of formatting, which are not in-line with the style used by the rest of the codebase. Once these changes are reverted, I think this PR would be good to go.

Of course ! Sorry about that. It is probably my black config.
I will fix it.

**Pull request closed without merging**